File: static/python/average_top_3_scores.py
import sys
import operator
import collections
import random
import string
import heapq


# @include
def find_student_with_highest_best_of_three_scores(name_score_data):
    student_scores = collections.defaultdict(list)
    for line in name_score_data:
        name, score = line.split()
        if len(student_scores[name]) < 3:
            heapq.heappush(student_scores[name], int(score))
        else:
            heapq.heappushpop(student_scores[name], int(score))
    return max([(sum(scores), name) for name, scores in student_scores.items()
                if len(scores) == 3],
               key=operator.itemgetter(0),
               default='no such student')[1]
# @exclude


def rand_string(length):
    return ''.join(random.choice(string.ascii_lowercase) for _ in range(length))


def simple_test():
    with open('scores.txt', 'w') as ofs:
        ofs.write('''adnan 100
amit 99
adnan 98
thl 90
adnan 10
amit 100
thl 99
thl 95
dd 100
dd 100
adnan 95''')
    with open('scores.txt') as name_score_data:
        result = find_student_with_highest_best_of_three_scores(name_score_data)
    print('result =', result)
    assert result == 'adnan'


def main():
    simple_test()
    n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 10000)

    with open('scores.txt', 'w') as ofs:
        for i in range(n):
            test_num = random.randint(0, 20)
            name = rand_string(random.randint(5, 10))
            for _ in range(test_num):
                print(name, random.randint(0, 100), file=ofs)
    with open('scores.txt') as name_score_data:
        name = find_student_with_highest_best_of_three_scores(name_score_data)
        name_score_data.seek(0)
    print('top student is', name)


if __name__ == '__main__':
    main()

File: static/cpp/average-top-3-scores.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <bitset>
#include <cassert>
#include <fstream>
#include <iostream>
#include <iterator>
#include <map>
#include <queue>
#include <random>
#include <set>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

using std::advance;
using std::bitset;
using std::cout;
using std::default_random_engine;
using std::endl;
using std::greater;
using std::ifstream;
using std::invalid_argument;
using std::ios;
using std::map;
using std::multiset;
using std::ofstream;
using std::priority_queue;
using std::random_device;
using std::stoi;
using std::string;
using std::uniform_int_distribution;
using std::unordered_map;
using std::vector;

int GetTopThreeScoresSum(priority_queue<int, vector<int>, greater<>>);

// @include
string FindStudentWithHighestBestOfThreeScores(ifstream* ifs) {
  // Use a multiset to handle duplicated test scores.
  unordered_map<string, priority_queue<int, vector<int>, greater<>>>
      student_scores;
  string name;
  int score;
  while (*ifs >> name >> score) {
    student_scores[name].emplace(score);
    if (student_scores[name].size() > 3) {
      student_scores[name].pop();  // Only keep the top 3 scores.
    }
  }

  string top_student = "no such student";
  int current_top_three_scores_sum = 0;
  for (const auto& scores : student_scores) {
    if (scores.second.size() == 3) {
      int current_scores_sum = GetTopThreeScoresSum(scores.second);
      if (current_scores_sum > current_top_three_scores_sum) {
        current_top_three_scores_sum = current_scores_sum;
        top_student = scores.first;
      }
    }
  }
  return top_student;
}

// Returns the sum of top three scores.
int GetTopThreeScoresSum(priority_queue<int, vector<int>, greater<>> scores) {
  int sum = 0;
  while (!scores.empty()) {
    sum += scores.top();
    scores.pop();
  }
  return sum;
}
// @exclude

string RandString(int len) {
  default_random_engine gen((random_device())());
  uniform_int_distribution<int> dis('a', 'z');
  string ret;
  while (len--) {
    ret += dis(gen);
  }
  return ret;
}

void SimpleTest() {
  ofstream ofs("scores.txt");
  ofs << "adnan 100";
  ofs << "amit 99";
  ofs << "adnan 98";
  ofs << "thl 90";
  ofs << "adnan 10";
  ofs << "amit 100";
  ofs << "thl 99";
  ofs << "thl 95";
  ofs << "adnan 95";
  ofs.close();
  ifstream ifs("scores.txt");
  string result = FindStudentWithHighestBestOfThreeScores(&ifs);
  cout << "result = " << result << endl;
  assert(result == "adnan");
}

int main(int argc, char* argv[]) {
  SimpleTest();
  default_random_engine gen((random_device())());
  int n;
  if (argc == 2) {
    n = stoi(argv[1]);
  } else {
    uniform_int_distribution<int> dis(1, 10000);
    n = dis(gen);
  }
  ofstream ofs("/tmp/scores.txt");
  for (int i = 0; i < n; ++i) {
    uniform_int_distribution<int> test_num_dis(0, 20);
    int test_num = test_num_dis(gen);
    uniform_int_distribution<int> len_dis(5, 10);
    string name = RandString(len_dis(gen));
    while (test_num--) {
      uniform_int_distribution<int> score_dis(0, 100);
      ofs << name << " " << score_dis(gen) << endl;
    }
  }
  ofs.close();
  ifstream ifs("/tmp/scores.txt");
  string name = FindStudentWithHighestBestOfThreeScores(&ifs);
  cout << "top student is " << name << endl;
  // Remove file after the execution.
  // remove("score.txt");
  return 0;
}
